src/search3_jobs.erl

```diff
@@ -0,0 +1,86 @@
+-module(search3_jobs).
+
+-export([
+    set_timeout/0,
+    build_search/3,
+    build_search_async/2
+]).
+
+-include("search3.hrl").
+
+set_timeout() ->
+    couch_jobs:set_type_timeout(?SEARCH_JOB_TYPE, 6 * 1000).
+
+build_search(TxDb, Index, UpdateSeq) ->
+    {ok, JobId} = build_search_async(TxDb, Index),
+    case wait_for_job(JobId, UpdateSeq) of
+        {ok, Session} -> Session;
+        retry -> build_search(TxDb, Index, UpdateSeq)
+    end.
+
+build_search_async(TxDb, Index) ->
+    JobId = job_id(TxDb, Index),
+    JobData = job_data(TxDb, Index),
+    ok = couch_jobs:add(undefined, ?SEARCH_JOB_TYPE, JobId, JobData),
+    {ok, JobId}.
+
+wait_for_job(JobId, UpdateSeq) ->
+    case couch_jobs:subscribe(?SEARCH_JOB_TYPE, JobId) of
+        {ok, Subscription, _State, _Data} ->
+            wait_for_job(JobId, Subscription, UpdateSeq);
+        {ok, finished, Data} ->
+            case Data of
+                #{<<"search_seq">> := SearchSeq, <<"session">> := Session}
+                        when SearchSeq >= UpdateSeq ->
+                    {ok, Session};
+                _ ->
+                    retry
+            end
+    end.
+
+wait_for_job(JobId, Subscription, UpdateSeq) ->
+    case wait(Subscription) of
+        {error, Error} ->
+            erlang:error(Error);
+        {finished, #{<<"error">> := Error, <<"reason">> := Reason}} ->
+            erlang:error({binary_to_atom(Error, latin1), Reason});
```

Comment:
this risks blowing up the atom table. if binary_to_existing_atom would not suffice, this is a blocker.